### ISSUE TITLE
SpnegoHandshake fails when more than one round is needed

### ISSUE THREAD

**Issue by matthiasblaesing** [AgNO3/jcifs-ng#10]
My test-setup:

I have a Windows 2016 Server Evaluation, installed as Domain Controller. That VM is equipped with three virtual nics (NAT for Internet Connection, Host-Only Networking for Debugging and and internal network). I have a second Windows 2016 server, that is joined to the domain. This setup was intended for experiments with Kerberos.

I wrote a minimal GSS<->SSPI Adapter, that exposes the windows SSPI functions wrapped into GSS Context/Credentials. Background is, that microsoft has shutdown the access to the TGT, that Oracle/OpenJDK java need to implement Kerberos Authentification.

While I tested jcifs-ng I noticed an odd behavior on the domain controller. On the secondary server I can repeatedly open connections and do listings, on the domain controller only the first connection attempt works. The second fails:

```
Jun 20, 2017 10:35:55 PM jcifs.smb.SmbSessionImpl sessionSetupSMB1
WARNUNG: Exception during SSP authentication
jcifs.smb.SmbException: Server chose an unsupported mechanism null
        at jcifs.smb.SpnegoContext.negotitate(SpnegoContext.java:193)
        at jcifs.smb.SpnegoContext.initSecContext(SpnegoContext.java:174)
        at jcifs.smb.SmbSessionImpl$4.run(SmbSessionImpl.java:822)
        at jcifs.smb.SmbSessionImpl$4.run(SmbSessionImpl.java:818)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:422)
        at jcifs.smb.SmbSessionImpl.sessionSetupSMB1(SmbSessionImpl.java:818)
        at jcifs.smb.SmbSessionImpl.sessionSetup(SmbSessionImpl.java:466)
        at jcifs.smb.SmbSessionImpl.send(SmbSessionImpl.java:358)
```

I looked into it further and modified the `jcifs.smb.SpnegoContext.negotitate` method in line 192 not to check of supported mechanism. (lines 192-194 were commented) And this basicly fixes the issue. 

Here you find a trace for both cases (pcap file in ZIP):

[successful_and_unsuccessfull_cap_2.pcap.zip](https://github.com/AgNO3/jcifs-ng/files/1089648/successful_and_unsuccessfull_cap_2.pcap.zip)

The first handshake is the expected sequence:

- The GSS Context is established on the client and the initial token is generated, SSPI indicates `SEC_I_CONTINUE_NEEDED`
- The token is send to the server (Packet number 49) and the server replies with the reply token (Packet number 51)
- The reply token is fed into SSPI and that indicates 'SEC_E_OK' (the handshake is done)

The second is the one, that blows with the current implementation (the trace shows the working state!):

- The GSS Context is established on the client and the initial token is generated, SSPI indicates `SEC_I_CONTINUE_NEEDED`
- The token is send to the server (Packet number 113) and the server replies with the reply token (Packet number 115). The reply token indicates KRB5KRB_AP_ERR_MODIFIED
- The reply token is fed into SSPI and that indicates 'SEC_I_CONTINUE_NEEDED'
- A new token is generated and send to the server (Packet number 119) and now an accept reply token (Packet number 121) is generated
- The reply token is fed into SSPI and that indicates 'SEC_E_OK' (the handshake is done)

Looking at the network trace, I conclude, that the exception is generated, because the second round (Packet 121) does not indicate a supportedMech, the supportedMech was present in Packet number 119.

Long story short - my fix currently looks like this:

```patch
--- a/src/main/java/jcifs/smb/SpnegoContext.java
+++ b/src/main/java/jcifs/smb/SpnegoContext.java
@@ -189,7 +189,7 @@
 
             if ( spToken instanceof NegTokenTarg ) {
                 NegTokenTarg targ = (NegTokenTarg) spToken;
-                if ( !this.mechContext.isSupported(targ.getMechanism()) ) {
+                if ( targ.getMechanism() != null && !this.mechContext.isSupported(targ.getMechanism()) ) {
                     throw new SmbException("Server chose an unsupported mechanism " + targ.getMechanism());
                 }
             }
```

I did not yet dive into the negotiate protocol, so this is most probably to easy.

**Comment by mbechler** [comment]
From the SPNEGO side of things this looks like legal behavior
```
  supportedMech

      This field SHALL only be present in the first reply from the
      target.  It MUST be one of the mechanism(s) offered by the
      initiator.
```
and
```
  -- REQUIRED in the first reply from the target
 supportedMech   [1] MechType      OPTIONAL,
```
so I guess we should still do that check for the first reply we receive in a negotiation, probably save the mechanism.

From the kerberos side this looks a bit strange. The server does reject the AP req, which suggests that something is wrong with it (all the plaintext parts look the same to me). Given that reply I would expect the GSSAPI/SSPI impl to err out, but it doesn't. Instead it "retries" (modulo some modifications in the encrypted parts) the request which then succeeds. 

If that works with SSPI (I'm pretty sure it won't with other GSSAPI impls), that's fine with me, still would be curious what kind of magic is going on there. Is there some related KDC traffic? Maybe kerberos event logging can also shed some light.

**Comment by matthiasblaesing** [comment]
I ran your change through my test and it works. Thanks for taking care of this. On optimisation, that might be interesting for the future: If anyone is interested in the negotiated mech, instead of a firstResponse flag, you could store the selected mech:

```diff
--- a/src/main/java/jcifs/smb/SpnegoContext.java
+++ b/src/main/java/jcifs/smb/SpnegoContext.java
@@ -54,7 +54,7 @@
 
     private SSPContext mechContext;
     private Oid[] mechs;
-    private boolean firstResponse;
+    private Oid negotiatedMech;
 
 
     /**
@@ -188,12 +188,12 @@
         try {
             SpnegoToken spToken = getToken(inputBuf, offset, len);
 
-            if ( this.firstResponse && spToken instanceof NegTokenTarg ) {
+            if ( negotiatedMech == null && spToken instanceof NegTokenTarg ) {
                 NegTokenTarg targ = (NegTokenTarg) spToken;
                 if ( !this.mechContext.isSupported(targ.getMechanism()) ) {
                     throw new SmbException("Server chose an unsupported mechanism " + targ.getMechanism());
                 }
-                this.firstResponse = false;
+                this.negotiatedMech = targ.getMechanism();
             }
 
             Oid currentMech = null;
```

To the "correctness" aspect: If I read the SPNEGO protocol correctly, there are only two options how the handshake can end: Either it is rejected or accepted. In this case the server side indicates accept-incomplete, so the handshake has to go on (this behavior matches the descriptions for initSecContext (GSS) and InitializeSecurityContext (SSPI)).

Don't take this wrong: I also think this is strange, but it happens on a domain controller with multiple NICs. I can imagine some thinks going wrong/different in this case. My code in this exchange is basicly passing tokens to/from native code, so at this point I'll live with this double-handshake.
